File: Data Structures & Algorithms/palindromic-substrings/submission-1.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int n = s.size();

        // 中心擴展法
        int ans = 0;
        // 奇回文串
        for (int i = 0; i < n; i++) {
            int l = i, r = i;
            while (l >= 0 && r < n && s[l] == s[r]) {
                l--;
                r++;
                ans++;
            }
        } // 偶回文串
        for (int i = 0; i < n - 1; i++) {
            int l = i, r = i + 1;
            while (l >= 0 && r < n && s[l] == s[r]) {
                l--;
                r++;
                ans++;
            }
        }
        return ans;
    }
};

File: Data Structures & Algorithms/count-connected-components/submission-1.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        # first construct the adj list to access the neighbor info
        g = [[] for _ in range(n)]
        for x, y in edges:
            g[x].append(y)
            g[y].append(x)

        # use dfs to mark the same cc
        ans = 0
        vis = [False] * n
        def dfs(i: int) -> None:
            vis[i] = True
            for x in g[i]:
                if not vis[x]:
                    dfs(x)
        
            
        for i in range(n):
            if not vis[i]:
                dfs(i)
                ans += 1
        return ans
